libraries/AP_MSP/AP_MSP.h

```diff
@@ -0,0 +1,144 @@
+
+#define MSP_OSD_START 2048
+#define MSP_OSD_STEP_X 1
+#define MSP_OSD_STEP_Y 32
+#define MSP_OSD_POS(osd_setting) (MSP_OSD_START + osd_setting->xpos*MSP_OSD_STEP_X + osd_setting->ypos*MSP_OSD_STEP_Y)
+
+using namespace MSP;
+
+class AP_MSP
+{
+public:
+    AP_MSP();
+
+    ~AP_MSP();
+
+    /* Do not allow copies */
+    AP_MSP(const AP_MSP &other) = delete;
+    AP_MSP &operator=(const AP_MSP&) = delete;
+
+    // User settable parameters
+    static const struct AP_Param::GroupInfo var_info[];
+
+    // init - perform required initialisation
+    bool init();
+    void msp_telemetry_callback(uint16_t cmd_msp, uint8_t protocol_instance);
+
+    static AP_MSP *get_singleton(void)
+    {
+        return singleton;
+    }
+
+private:
+    // to access the telemetry_info member variable
+    friend class AP_MSP_Telem;
+    friend class AP_MSP_Telem_DJI;
+    friend class AP_MSP_Telem_Backend;
+
+    enum : uint8_t {
+        BACKEND_MSP,
+        BACKEND_DJI,
+        BACKEND_COUNT
+    };
+
+    enum {
+        OPTION_TELEMETRY_MODE = 1U<<0,
+        OPTION_INVERTED_WIND = 1U<<1
+    };
+
+    AP_Int32 options;
+
+    AP_Int8 units;
+    AP_Int8 msgtime_s;
+    AP_Int8 cellcount;
+
+    // these are the osd items we support for MSP OSD
+    AP_OSD_Setting* osd_item_settings[58] {};
+
+    const int16_t osd_enabled_stats[24] = {
+        -1,                         // OSD_STAT_RTC_DATE_TIME
+            -1,                         // OSD_STAT_TIMER_1
+            -1,                         // OSD_STAT_TIMER_2
+            -1,                         // OSD_STAT_MAX_SPEED
+            -1,                         // OSD_STAT_MAX_DISTANCE
+            -1,                         // OSD_STAT_MIN_BATTERY
+            -1,                         // OSD_STAT_END_BATTERY
+            -1,                         // OSD_STAT_BATTERY
+            -1,                         // OSD_STAT_MIN_RSSI
+            -1,                         // OSD_STAT_MAX_CURRENT
+            -1,                         // OSD_STAT_USED_MAH
+            -1,                         // OSD_STAT_MAX_ALTITUDE
+            -1,                         // OSD_STAT_BLACKBOX
+            -1,                         // OSD_STAT_BLACKBOX_NUMBER
+            -1,                         // OSD_STAT_MAX_G_FORCE
+            -1,                         // OSD_STAT_MAX_ESC_TEMP
+            -1,                         // OSD_STAT_MAX_ESC_RPM
+            -1,                         // OSD_STAT_MIN_LINK_QUALITY
+            -1,                         // OSD_STAT_FLIGHT_DISTANCE
+            -1,                         // OSD_STAT_MAX_FFT
+            -1,                         // OSD_STAT_TOTAL_FLIGHTS
+            -1,                         // OSD_STAT_TOTAL_TIME
+            -1,                         // OSD_STAT_TOTAL_DIST
+            -1,                         // OSD_STAT_MIN_RSSI_DBM
+        };
+
+    AP_MSP_Telem_Backend::telemetry_info_t telemetry_info;
```

Comment:
I think convention for these type names is camel-case?
Some comments would be nice

Reply:
I'm going to remove this one, camel case for types with _ separator like AP_Camel_Case?

Reply:
For internal types it tends to be MyTypeThatILike. For AP classes AP_AnotherType